fairlearn/metrics/_make_derived_metric.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) Microsoft Corporation and Fairlearn contributors.
+# Licensed under the MIT License.
+
+import functools
+import inspect
+from typing import Callable, List, Union
+
+from ._metric_frame import MetricFrame
+
+transform_options = [
+    'difference',
+    'group_min',
+    'group_max',
+    'ratio',
+]
+
+parameters_for_transforms = [
+    'method'
+]
+
+_METRIC_CALLABLE_ERROR = "Supplied metric object must be callable"
+_METHOD_ARG_ERROR = "Callables which accept a '{0}' argument " \
+    "may not be passed to make_derived_metric(). Please use functools.partial()"
+_INVALID_TRANSFORM = "Transform must be one of {0}".format(transform_options)
+
+
+class _DerivedMetric:
+    def __init__(self,
+                 *,
+                 metric: Callable[..., Union[float, int]],
+                 transform: str,
+                 sample_param_names: List[str]):
+
+        if not callable(metric):
+            raise ValueError(_METRIC_CALLABLE_ERROR)
+        sig = inspect.signature(metric)
+
+        for param_name in parameters_for_transforms:
+            if param_name in sig.parameters:
+                raise ValueError(_METHOD_ARG_ERROR.format(param_name))
+        self._metric_fn = metric
+
+        if transform not in transform_options:
+            raise ValueError(_INVALID_TRANSFORM)
+        self._transform = transform
+
+        self._sample_param_names = []
+        if sample_param_names is not None:
+            self._sample_param_names = sample_param_names
+
+    def __call__(self,
+                 y_true,
+                 y_pred,
+                 *,
+                 sensitive_features,
+                 **other_params) -> Union[float, int]:
+        sample_params = dict()
+        params = dict()
+        transform_parameters = dict()
+        for k, v in other_params.items():
+            if k in self._sample_param_names:
+                sample_params[k] = v
+            elif k in parameters_for_transforms:
+                transform_parameters[k] = v
+            else:
+                params[k] = v
+
+        dispatch_fn = functools.partial(self._metric_fn, **params)
+        # Make sure there isn't a subsequent log message about
+        # a nameless metric
+        bound_fn_name = self._metric_fn.__name__
+        for k, v in params.items():
```

Comment:
you may want to sort the keys before getting to this loop (`sorted(params.items())`), to get a deterministic fn name independent of insertion order to the params dict: https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6

Reply:
Good point - thanks for the reminder.